geoopt/manifolds/birkhoff_polytope.py

```diff
@@ -0,0 +1,143 @@
+import torch
+
+from geoopt import linalg
+from .base import Manifold
+
+__all__ = ["BirkhoffPolytope"]
+
+_birkhoff_doc = r"""
+    Manifold induced by the Doubly Stochastic matrices as described in 
+    A. Douik and B. Hassibi, "Manifold Optimization Over the Set 
+    of Doubly Stochastic Matrices: A Second-Order Geometry"
+    ArXiv:1802.02628, 2018.
+    Link to the paper: https://arxiv.org/abs/1802.02628.
+
+    @Techreport{Douik2018Manifold,
+       Title   = {Manifold Optimization Over the Set of Doubly Stochastic 
+                  Matrices: {A} Second-Order Geometry},
+       Author  = {Douik, A. and Hassibi, B.},
+       Journal = {Arxiv preprint ArXiv:1802.02628},
+       Year    = {2018}
+    }
+    
+    Please also cite:
+    Tolga Birdal, Umut Şimşekli,
+    "Probabilistic Permutation Synchronization using the Riemannian Structure of the BirkhoffPolytope Polytope"
+    IEEE Conference on Computer Vision and Pattern Recognition, CVPR, 2019
+    Link to the paper: https://arxiv.org/abs/1904.05814
+"""
+
+
+class BirkhoffPolytope(Manifold):
+    __doc__ = r"""
+    {}
+    """.format(_birkhoff_doc)
+    name = "BirkhoffPolytope"
+
+    ## batch_size \times n \times n
+    ndim = 3
+
+    def __init__(self, maxiter=100, tol=1e-3, epsilon=1e-12):
+        self.maxiter = maxiter
+        self.tol = tol
+        self.epsilon = epsilon
+
+    def _check_shape(self, x, name):
+        dim_is_ok = x.dim() == self.ndim
+        if not dim_is_ok:
+            return False, "dimensions for `{}` not match".format(name)
+
+        return True, None
+
+    def _check_point_on_manifold(self, x, atol=1e-5, rtol=1e-5):
+        row_sum = x.sum(dim=-1)
+        col_sum = x.sum(dim=-2)
+        row_ok = torch.allclose(row_sum, row_sum.new((1,)).fill_(1), atol=atol, rtol=rtol)
+        col_ok = torch.allclose(col_sum, col_sum.new((1,)).fill_(1), atol=atol, rtol=rtol)
+        if row_ok and col_ok:
+            return True, None
+        else:
+            return False, "illegal doubly stochastic matrix with atol={}, rtol={}".format(atol, rtol)
+
+    def _check_vector_on_tangent(self, x, u, atol=1e-5, rtol=1e-5):
+        diff = u.transpose(-1, -2) @ x + x.transpose(-1, -2) @ u
+        ok = torch.allclose(diff, diff.new((1,)).fill_(0), atol=atol, rtol=rtol)
+        if not ok:
+            return False, "`u^T x + x^T u !=0` with atol={}, rtol={}".format(atol, rtol)
+        return True, None
+
+    def _projx(self, x):
+        iter = 0
+        x_shape = x.shape
+        x = x.reshape(-1, x_shape[-2], x_shape[-1])
+        c = 1.0 / (torch.sum(x, dim=-2, keepdim=True) + self.epsilon)
+        r = 1.0 / (torch.matmul(x, torch.transpose(c, -1, -2)) + self.epsilon)
+        while iter < self.maxiter:
+            iter += 1
+            cinv = torch.matmul(torch.Tensor.permute(r, 0, 2, 1), x)
```

Comment:
```suggestion
            cinv = torch.matmul(r.permute(0, 2, 1), x)
```

This might be more explicit